notificationdemo/source/java/src/org/kaaproject/kaa/demo/notification/NotificationDemo.java

```diff
@@ -51,74 +58,107 @@ public static void main(String[] args) {
             }
         });
 
         // Start the Kaa client and connect it to the Kaa server.
         kaaClient.start();
 
         // Get available notification topics.
-        List<Topic> topicList = kaaClient.getTopics();
+        topics = kaaClient.getTopics();
 
         // List the obtained notification topics.
-        showTopicList(topicList);
-
-        try {
-            
-            // Wait for some input before exiting.
-            System.in.read();
-        } catch (IOException e) {
-            LOG.error("IOException was caught", e);
+        showTopics();
+
+        LOG.info("Type topic ID in order to subscribe to ones or type any text to exit");
+        Scanner scanner = new Scanner(System.in);
+        while (scanner.hasNextLong()) {
+            long topicId = scanner.nextLong();
+            if (getTopic(topicId) != null) {
+                LOG.info("Subscribing to optional topic {}", topicId);
+                subscribeTopic(topicId);
+            } else {
+                LOG.info("There is no input topic id. Please, input existing topic id.");
+            }
         }
 
         // Stop listening to the notification topic list updates.
         kaaClient.removeTopicListListener(topicListListener);
+        unsubscribeOptionalTopics();
 
         // Stop the Kaa client and release all the resources which were in use.
         kaaClient.stop();
         LOG.info("Notification demo stopped");
     }
 
-    // A listener that tracks the notification topic list updates
-    // and subscribes the Kaa client to every new topic available.
-    private static class BasicNotificationTopicListListener implements NotificationTopicListListener {
-        @Override
-        public void onListUpdated(List<Topic> list) {
-            LOG.info("Topic list was updated");
-            showTopicList(list);
-            try {
-                //Try to subscribe to all new optional topics, if any.
-                List<Long> optionalTopics = extractOptionalTopicIds(list);
-                for(Long optionalTopicId : optionalTopics){
-                    LOG.info("Subscribing to optional topic {}", optionalTopicId);
-                }
-                kaaClient.subscribeToTopics(optionalTopics, true);
-            } catch (UnavailableTopicException e) {
-                LOG.error("Topic is unavailable, can't subscribe: {}", e.getMessage());
-            }
+    private static void showTopics() {
+        if (topics == null || topics.isEmpty()) {
+            LOG.info("Topic list is empty");
+            return;
+        }
+
+        LOG.info("Available topics:");
+        for (Topic topic : topics) {
+            LOG.info("Topic id: {}, name: {}, type: {}", topic.getId(), topic.getName(), topic.getSubscriptionType());
+        }
+
+        LOG.info("Subscribed on topics:");
+        for (Topic t : getOneTypeTopics(SubscriptionType.MANDATORY_SUBSCRIPTION)) {
```

Comment:
I propose to add one more list in which add IDs of notification topics that endpoint is subscribed on. And dump this list into file between demo app launches, ie if this file exists in **application** working directory fetch list from the file and if not - create one and save the topics IDs into it on application shutdown. (Unfortunately, we can't get the list of subscriptions on a client side and we are forced to "memorize" it)

The user will be confused if we will show in the list of subscriptions only mandatory topics.